dll/win32/shell32/wine/shellpath.c

```diff
@@ -474,42 +595,118 @@ int WINAPI PathCleanupSpec( LPCWSTR lpszPathW, LPWSTR lpszFileW )
 }
 
 /*************************************************************************
- * PathQualifyA		[SHELL32]
+ * PathQualifyW		[SHELL32]
  */
-static BOOL PathQualifyA(LPCSTR pszPath)
+static VOID PathQualifyW(LPWSTR pszPath)
 {
-	FIXME("%s\n",pszPath);
-	return FALSE;
+    TRACE("%s\n",debugstr_w(pszPath));
+    PathQualifyExW(pszPath, NULL, 0);
 }
 
 /*************************************************************************
- * PathQualifyW		[SHELL32]
+ * PathQualifyA		[SHELL32]
  */
-static BOOL PathQualifyW(LPCWSTR pszPath)
+static VOID PathQualifyA(LPSTR pszPath)
 {
-	FIXME("%s\n",debugstr_w(pszPath));
-	return FALSE;
+    WCHAR szPath[MAX_PATH];
+    TRACE("%s\n",pszPath);
+    SHAnsiToUnicode(pszPath, szPath, _countof(szPath));
+    PathQualifyW(szPath);
+    SHUnicodeToAnsi(szPath, pszPath, MAX_PATH);
 }
 
 /*************************************************************************
  * PathQualify	[SHELL32.49]
  */
-BOOL WINAPI PathQualifyAW(LPCVOID pszPath)
+VOID WINAPI PathQualifyAW(LPVOID pszPath)
 {
-	if (SHELL_OsIsUnicode())
-	  return PathQualifyW(pszPath);
-	return PathQualifyA(pszPath);
+    if (SHELL_OsIsUnicode())
+        PathQualifyW(pszPath);
+    else
+        PathQualifyA(pszPath);
 }
 
-static BOOL PathResolveA(LPSTR path, LPCSTR *paths, DWORD flags)
+BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
 {
-    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_a(path), paths, flags);
+    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_a(path), dirs, flags);
     return FALSE;
 }
 
-static BOOL PathResolveW(LPWSTR path, LPCWSTR *paths, DWORD flags)
+BOOL WINAPI PathResolveW(LPWSTR path, LPCWSTR *dirs, DWORD flags)
 {
-    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_w(path), paths, flags);
+    DWORD dwWhich;
+    TRACE("PathResolveW(%s,%p,0x%08x)\n", debugstr_w(path), dirs, flags);
+
+    dwWhich = ((flags & PRF_DONTFINDLNK) ? 0xF : 0x3F);
+
+    if (flags & PRF_VERIFYEXISTS)
+        SetLastError(ERROR_FILE_NOT_FOUND);
+
+    PathUnquoteSpacesW(path);
+
+    if (PathIsRootW(path))
+    {
+        if ((path[0] == L'\\' && path[1] == 0) ||
+            PathIsUNCServerW(path) || PathIsUNCServerShareW(path))
+        {
+            if (flags & PRF_FIRSTDIRDEF)
+                PathQualifyExW(path, dirs[0], 0);
+            else
+                PathQualifyExW(path, NULL, 0);
+        }
+
+        if (flags & PRF_VERIFYEXISTS)
+            return PathFileExistsAndAttributesW(path, NULL);
+        return TRUE;
+    }
+    else if (PathIsFileSpecW(path))
+    {
+        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))
+            return TRUE;
+
+        if (PathFindOnPathW(path, dirs))
+        {
+#ifdef REQUIREABSOLUTE
```

Comment:
Why is there a define for that? Can you just compile unconditionally the support for PRF_REQUIREABSOLUTE?

Reply:
We have to support `PRF_REQUIREABSOLUTE` in the future.

Reply:
Sure, then just compile it without that ifdef 🤷 
Or is it something that didn't exist in Windows 2003, but appeared in Vista+ ?

Reply:
`PRF_REQUIREABSOLUTE` doesn't work on normal status of Win2k3 at least.

Reply:
I will delete PRF_REQUIREABSOLUTE impl.

Reply:
nono, you can just keep it for the time being, but maybe surrounded between `#if (NTDDI_VERSION >= NTDDI_VISTA)  ...   #endif`

Reply:
OK.

Reply:
`_WIN32_WINNT` and `NTDDI_VERSION` are mismatching...

Reply:
ok use whatever define we use in shell32.

Reply:
`_WIN32_WINNT=0x502` stands for `_WIN32_WINNT_WS03` (Server 2003).
`_WIN32_WINNT=0x600` stands for `_WIN32_WINNT_WS08` (Vista+).
We are using `_WIN32_WINNT_WS08` in `CMakeLists.txt`.

Reply:
and this is also `_WIN32_WINNT_VISTA`.

Reply:
Should I delete this?
```txt
remove_definitions(-D_WIN32_WINNT=0x502)
add_definitions(-D_WIN32_WINNT=0x600)
```

Reply:
Please keep these, but delete the NTDDI thing you've added.

Reply:
> Deleted.

I wasn't asking to remove the implementation of `PRF_REQUIREABSOLUTE`, but merely to remove the "NTDDI" line you added in the CMakeLists.txt file, and keeping the `D_WIN32_WINNT` things inside this file, and modifying the `#ifdefs` around the `PRF_REQUIREABSOLUTE` support to use the `_WIN32_WINNT` instead... 🙄

Reply:
OK.